exonum/src/blockchain/service/srv.rs

```diff
@@ -0,0 +1,247 @@
+
+
+use super::{Transaction, ServiceContext, ApiContext, Service};
+use std::marker::PhantomData;
+
+/// TODO
+pub trait TransactionSet {
+    /// TODO
+    fn tx_from_raw(raw: RawTransaction) -> Result<Box<Transaction>, MessageError>;
+    /// TODO
+    fn tx_from_request(request: &mut Request) -> Result<Box<Transaction>, ApiError>;
+}
+
+#[macro_export]
+macro_rules! transaction_set {
+    ( $name:ident { $($tx:ident),* } ) => {
+        pub enum $name {
+        }
+
+        impl $crate::blockchain::TransactionSet for $name {
+            fn tx_from_raw(raw: RawTransaction) -> Result<Box<$crate::blockchain::Transaction>, $crate::encoding::Error> {
+                let message_type = raw.message_type();
+                $(
+                    if $tx::message_id() == message_type {
+                        return Ok(Box::new($tx::from_raw(raw)?));
+                    }
+                )*
+
+                return Err($crate::encoding::Error::IncorrectMessageType { message_type })
+            }
+
+            //TODO: reexport & $crate::
+            fn tx_from_request(request: &mut $crate::iron::Request) -> Result<Box<$crate::blockchain::Transaction>, $crate::api::ApiError> {
+                use $crate::iron::prelude::*;
+
+                #[derive(Deserialize, Clone)]
+                #[serde(untagged)] // :(
+                enum Any {
+                    $($tx($tx),)*
+                }
+
+                match request.get::<$crate::bodyparser::Struct<Any>>() {
+                    Ok(None) => Err($crate::api::ApiError::IncorrectRequest("Empty request body".into()))?,
+                    Err(e) => Err($crate::api::ApiError::IncorrectRequest(Box::new(e)))?,
+                    Ok(Some(any)) => {
+                        match any {$(
+                            Any::$tx(tx) => Ok(Box::new(tx)),
+                        )*}
+                    }
+                }
+
+            }
+        }
+    }
+}
+
+/// The structure returned by the REST API.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TransactionResponse {
+    /// TODO
+    pub tx_hash: Hash,
+}
+
+
+/// Service capable of receiving and processing transactions
+pub trait TransactionService: Send + Sync + 'static {
+    /// TODO
+    type Transactions: TransactionSet;
+    // = NoTransactions
+    /// TODO
+    const ID: u16;
+    /// TODO
+    const NAME: &'static str;
+
+    /// TODOk
+    fn state_hash(&self, snapshot: &Snapshot) -> Vec<Hash>;
+
+    /// TODO
+    fn initialize(&self, _fork: &mut Fork) -> Value {
+        Value::Null
+    }
+
+    /// TODO
+    fn handle_commit(&self, _context: &ServiceContext) {
+    }
+
+    /// TODO
+    fn wire_public_api(&self, _router: &mut Router, _ctx: &ApiContext) {
+    }
+
+    /// TODO
+    fn wire_private_api(&self, _router: &mut Router, _ctx: &ApiContext) {
+    }
+
+    /// TODO
+    fn into_service(self) -> Box<Service> where Self: Sized {
+        Box::new(TransactionServiceImpl(self))
+    }
```

Comment:
This is an interesting method. Ideally, we'd like to be able to define  `impl<S: TransactionService> Into<Box<Service>> for S`, but it's possible to have only one such impl, which makes defining several base traits impossible. To put it in other way, nothing prevents a user from implementing *both* `TransactionService` and `ObserverService` for a single type, so we can't generically convert them to a `Box<Service>`.